### DISCUSSION TITLE
CustomTextEditor vs regular webview

### DISCUSSION THREAD

**Discussion by Andarist** [microsoft/vscode-discussions#89] [Extension Development QnA] [answered]
A regular webview can implement syncing between its local state and the text file. It also involves a similar level of complexity compared to dealing with CustomTextEditors (the CustomTextEditor doesn't seem to make it a lot easier to deal with bidirectional syncing). So I wonder - when I should prefer using the CustomTextEditor over a plain webview panel? What are strong reasons to use it? Are there any things that are considerably easier with it?

**Comment by isidorn** [comment, marked as answer]
A WebView gives you more flexibility, but a CustomTextEditor is better integrated into VS Code. So a rule of thumb if you can use a CustomTextEditor go ahead and use it. If some features are missing then you should go ahead and go for a webivew.

I will go ahead and mark this as answered, and if you have more questions do let us know. Thanks

**Reply by Andarist** [reply]
> A WebView gives you more flexibility, but a CustomTextEditor is better integrated into VS Code.

This is kinda what I've assumed **but** it's a very vague statement and I couldn't figure out from the docs alone what would be the actual benefit for me if I would switch from a webview to a CustomTextEditor

**Reply by r3m0t** [reply]
The TextDocument class handles the syncing between the filesystem and the `TextDocument.getText()`. Actions in VS Code like save, undo/redo, and `WorkspaceEdit` actions such as 'replace all in files' or 'amend import statements after file is renamed' can make changes to your `TextDocument`. Another feature handled is hot exit which is where unsaved content is shown in the editor after quitting vs code and reopening it.

Your extension needs to handle the display of `TextDocument.getText` in your webview and it needs to translate the user's changes into `WorkspaceEdit`s. You don't need to access the filesystem. Take a look at this example https://github.com/microsoft/vscode-extension-samples/blob/main/custom-editor-sample/src/catScratchEditor.ts

**Reply by Andarist** [reply]
> The TextDocument class handles the syncing between the filesystem and the TextDocument.getText()

The direct access to `getText()` is indeed nice, but it's just saving me a lookup for the correct document. From what I can tell `.getText()` is already synced correctly in the client part of the extension and you still need to pass it around to the webview manually. I can see in the mentioned example that it's just sent through `postMessage` here:
https://github.com/microsoft/vscode-extension-samples/blob/31d3d4920b3fae52e03a185e434f51ee81965515/custom-editor-sample/src/catScratchEditor.ts#L51

> Actions in VS Code like save, undo/redo, and WorkspaceEdit actions such as 'replace all in files' or 'amend import statements after file is renamed' can make changes to your TextDocument.

Right - I understand that different things can edit the text. But the linked example also has to listen for this stuff manually, like like I would do with a regular webview. It just listens to `onDidChangeTextDocument `:
https://github.com/microsoft/vscode-extension-samples/blob/31d3d4920b3fae52e03a185e434f51ee81965515/custom-editor-sample/src/catScratchEditor.ts#L63

> Another feature handled is hot exit which is where unsaved content is shown in the editor after quitting vs code and reopening it.

This one I see as an advantage, it feels like an added feature.

> Your extension needs to handle the display of TextDocument.getText in your webview and it needs to translate the user's changes into WorkspaceEdits.

This feels exactly the same as with regular webviews. I do exactly the same thing with a regular webview (create WorkspaceEdits) as a `CustomTextEditor` has to do:
https://github.com/microsoft/vscode-extension-samples/blob/31d3d4920b3fae52e03a185e434f51ee81965515/custom-editor-sample/src/catScratchEditor.ts#L194-L203

---

One thing I was hoping to get out of the `CustomTextEditor` was to get access to the undo command or smth like that - but I couldn't yet figure out how to call this.

**Reply by isidorn** [reply]
`CustomTextEditor` is handled more like an editor - that means that it has a `TextDocument` attached to it, and it has a `uri`. Because of that it can be dirty, will be saved when the user does a "save all", can have selection etc..

@lramos15 can provide more benefits if needed, and can clarify if it participates in the undo.

**Reply by lramos15** [reply]
@Andarist What it boils down to is a `CustomTextEditor` is a webview panel with a backing document who's data model is managed by core VS Code. So I would say the advantages like mentioned above are just more syntactic sugar on top of a normal webview panel. I don't know much about your extension but it sounds like you've implemented a lot of the stuff that `CustomTextEditors` provide for you. As for undo redo that works the same as core vs code. Whenever a worskpace edit is applied to the backing text document then we add to the undo / redo stack.
